### PR TITLE
Add txn-no as a pagination token into listaccounthistory rpc

### PR DESCRIPTION
**Pull request by chanakasameera**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Currently listaccounthistory on DeFiCh/ain. There isn't a way to effectively pagination as you can't provide txn-no as a pagination token. To support slice pagination; we need to be able to provide to all params within the AccountHistoryKey. Address & BlockHeight is currently supported but, txn-no isn't.
#### Which issue(s) does this PR fixes?:
https://github.com/DeFiCh/jellyfish/issues/1036

Fixes #

### CODE REVIEW

File: src/masternodes/rpc_accounts.cpp [hunk 1]
```diff
@@ -1089,6 +1097,10 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
             return false;
         }
 
+        if(txn != std::numeric_limits<uint32_t>::max() && txn != key.txn) {
+            return true;
+        }
+
```
Comment (bvbfan) [inline src/masternodes/rpc_accounts.cpp:1103]: Move it below category check, at line 1125

**Review by bvbfan — CHANGES_REQUESTED** [review]

**Review by bvbfan — APPROVED** [review]

File: src/masternodes/rpc_accounts.cpp [hunk 2]
```diff
@@ -1117,6 +1125,10 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
             return true;
         }
 
+        if(txn != std::numeric_limits<uint32_t>::max() && txn != key.txn) {
+            return true;
+        }
+
```
Comment (surangap) [inline src/masternodes/rpc_accounts.cpp:1131]: hmm. @chanakasameera two things `filtration` and `pagination` within the same RPC. need to be careful. 
https://github.com/DeFiCh/jellyfish/issues/1036 requirement is to address slice pagination.
This way it's acting like a filter that we don't want and hardly someone would use.

**Review by bvbfan — DISMISSED** [review]

File: src/masternodes/rpc_accounts.cpp [hunk 3]
```diff
@@ -1184,6 +1192,9 @@ UniValue listaccounthistory(const JSONRPCRequest& request) {
                 return txs.count(pwtx->GetHash()) || startBlock > index->nHeight || index->nHeight > maxBlockHeight;
             },
             [&](COutputEntry const & entry, CBlockIndex const * index, CWalletTx const * pwtx) {
+                if(txn != std::numeric_limits<uint32_t>::max() && txn != entry.vout) {
+                    return true;
+                }
```
Comment (surangap) [inline src/masternodes/rpc_accounts.cpp:1197]: do we need this? still feels like a `filteration`

Comment (bvbfan) [reply src/masternodes/rpc_accounts.cpp:1197]: Good point, It should be better to be
```
if (txn != std::numeric_limits<uint32_t>::max() && index->nHeight == maxBlockHeight && txn < entry.vout)
```

File: test/functional/rpc_listaccounthistory.py [hunk 4]
```diff
@@ -72,6 +73,28 @@ def run_test(self):
         for txs in results:
             assert(hasattr(txs['amounts'], '__len__') and (not isinstance(txs['amounts'], str)))
 
+        # list {"maxBlockHeight":103, "txn":1}, should list without blockheight = 103, txn=2. i.e without MintToken
+        results = self.nodes[0].listaccounthistory(collateral_a, {"maxBlockHeight":103, "txn":1})
+        #Note(surangap): the results here has 2 extra send receive entries with different txids. not sure where they come from.
```
Comment (surangap) [inline test/functional/rpc_listaccounthistory.py:78]: @bvbfan any idea where these extra entries have come from?

Comment (bvbfan) [reply test/functional/rpc_listaccounthistory.py:78]: `Auto Auth` is used to ensure auth on specific address.

**Review by bvbfan — DISMISSED** [review]

File: test/functional/rpc_listaccounthistory.py (hunk 4 above)
Comment (surangap) [inline test/functional/rpc_listaccounthistory.py:78]:
```suggestion
```

**Review by surangap — APPROVED** [review]
`txn` is an input parameter added in this PR. So there can't be an already existing use-case that uses `txn` input parameter at Cake side. If they also do not depend on `txn` value in the response, then should be good to go.

**Pull request merged**